File: 21-merge-two-sorted-lists/21-merge-two-sorted-lists.php
class Solution {

     function insert(&$x, &$head, &$tail) {
        
        if($head == null) 
            $head = $x;
        else 
            $tail->next = $x;
        
        $tail = $x;
        $x = $x->next;
        
    }
    
    function mergeTwoLists($list1, $list2) {
        if($list1 == null) return $list2;
        if($list2 == null) return $list1;
        
        $head = null;
        $tail = null;
        
        while($list1 and $list2) {
            if($list1->val <= $list2->val) 
                Solution::insert($list1, $head, $tail);
            else 
                Solution::insert($list2, $head, $tail);
        }
        
        if($list1 == null) $tail->next = $list2;
        if($list2 == null) $tail->next = $list1;
        
        return $head;
        
    }
}

File: 21-merge-two-sorted-lists/21-merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        
        auto head = new ListNode(INT_MIN);
        auto tail = head;
        
        while(list1 or list2) {
            auto &cur = (list2 == NULL or (list1 != NULL and list1 -> val <= list2 -> val)) ? list1 : list2; 
            tail -> next = cur;
            cur = cur -> next;
            tail = tail -> next;
        }
        
        return head -> next;
    }
};